File: packages/dotnet/ReadMe/Webhook.cs
using System;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using Microsoft.Extensions.Primitives;
using static System.Text.Encoding;

namespace ReadMe
{
  public static class Webhook
  {
    public static void Verify(string body, StringValues signature, string secret)
    {
      if (signature.Count == 0)
      {
        throw new Exception("Missing Signature");
      }

      var input = signature.ToString().Split(",").Select(item =>
      {
        var split = item.Trim().Split("=", 2);
        return new { Key = split[0], Value = split[1] };
      })
      .ToLookup(item => item.Key, item => item.Value);

      var time = input["t"].First();
      if (DateTime.Now.Subtract(DateTimeOffset.FromUnixTimeMilliseconds(long.Parse(time)).DateTime).Minutes > 30)
      {
        throw new Exception("Expired Signature");
      }

      var readmeSignature = input["v0"].First();
      var unsigned = time + "." + body;
      var hmac = HMAC.Create("HMACSHA256");
      hmac.Initialize();
      hmac.Key = UTF8.GetBytes(secret);
      var result = UTF8.GetBytes(unsigned);
      var hash = hmac.ComputeHash(result);

      var verifySignature = new StringBuilder(hash.Length * 2);
      foreach (byte b in hash)
      {
        verifySignature.AppendFormat("{0:x2}", b);
      }

      if (verifySignature.ToString() != readmeSignature)
      {
        throw new Exception("Invalid Signature");
      }
    }
  }
}

File: packages/ruby/lib/readme/webhook.rb
require 'readme/metrics'

module Readme
  class MissingSignatureError < ArgumentError
    def message
      'Missing Signature'
    end
  end

  class ExpiredSignatureError < RuntimeError
    def message
      'Expired Signature'
    end
  end

  class InvalidSignatureError < RuntimeError
    def message
      'Invalid Signature'
    end
  end

  class Webhook
    def self.verify(body, signature, secret)
      raise MissingSignatureError unless signature

      parsed = signature.split(',').each_with_object({ time: -1, readme_signature: '' }) do |item, accum|
        k, v = item.split('=')
        accum[:time] = v if k.eql? 't'
        accum[:readme_signature] = v if k.eql? 'v0'
      end

      # Make sure timestamp is recent to prevent replay attacks
      thirty_minutes = 30 * 60
      raise ExpiredSignatureError if Time.now.utc - Time.at(0, parsed[:time].to_i, :millisecond).utc > thirty_minutes

      # Verify the signature is valid
      unsigned = "#{parsed[:time]}.#{body}"
      mac = OpenSSL::HMAC.hexdigest('SHA256', secret, unsigned)
      raise InvalidSignatureError if mac != parsed[:readme_signature]
    end
  end
end
